Support for Java 16 Record feature

PabloNicolasDiaz opened the pull request:
This PR adds support for java 16 record feature (https://openjdk.org/jeps/395). Users now can use `JavaClass.isRecord()` to detect a record class.

Update : Fixed PR description. The Record feature was introduced as a preview feature in Java 14 (https://openjdk.org/jeps/359) and Java 15 (https://openjdk.org/jeps/384), and its final release was in Java 16 (https://openjdk.org/jeps/395)

garydgregory commented:
@PabloNicolasDiaz 
Thank you for your PR.
I fixed a whole set of whitespace issues.

1. Please rebase on Git master to pick up a new version of our Checkstyle configuration, then
2. Run `mvn` by itself to run all build checks or `mvn checkstyle:check' to detect the remaining whitespace issues in this PR
3. Fix whatever the above detects then push your changes.

Question: Does the new class `RecordComponentInfo` need to be public? The general idea is to make as little as possible public.

Missing: All new public and protected items need Javadoc, and for some new items this PR is missing the first sentence in the Javadoc.

TY!

garydgregory reviewed (changes requested):
See my comments.

garydgregory commented on src/main/java/org/apache/bcel/classfile/JavaClass.java:
```diff
@@ -906,4 +910,22 @@ public String toString() {
         return buf.toString();
     }
+
+    public boolean isRecord() {
```
Javadoc missing, the next version will be 6.9.0 since this is a PR for a new feature.

garydgregory commented on src/main/java/org/apache/bcel/classfile/Record.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    /**
+     * @param nameIndex Index in constant pool
```
Javadoc's first sentence is missing.

garydgregory commented on src/main/java/org/apache/bcel/classfile/Record.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    /**
+     * @return deep copy of this attribute
```
Javadoc's first sentence is missing.

garydgregory commented on src/main/java/org/apache/bcel/classfile/RecordComponentInfo.java:
```diff
@@ -0,0 +1,84 @@
+
+import org.apache.bcel.Const;
+
+public class RecordComponentInfo implements Cloneable, Node {
```
Must the class be public. Let's not make it `Cloneable` if we can. In general, we want to move away from Cloneable and toward serialization proxies instead.

garydgregory commented on src/main/java/org/apache/bcel/classfile/RecordComponentInfo.java:
```diff
@@ -0,0 +1,84 @@
+    private final Attribute[] attributes; 
+    private final ConstantPool constantPool;
+    
+    public RecordComponentInfo(int index, int descriptorIndex, Attribute[] attributes, DataInput input,
```
Javadoc missing.

garydgregory commented on src/main/java/org/apache/bcel/classfile/RecordComponentInfo.java:
```diff
@@ -0,0 +1,84 @@
+        this.constantPool=constantPool;
+    }
+
+    public void write(DataOutputStream file) throws IOException {
```
Javadoc missing.

garydgregory commented on src/main/java/org/apache/bcel/classfile/Visitor.java:
```diff
@@ -207,6 +207,14 @@ default void visitNestMembers(final NestMembers obj) {
     }
 
+    /**
+     * @since 6.5.0
```
The next version will be 6.9.0.

garydgregory commented on src/test/java/org/apache/bcel/CounterVisitorTestCase.java:
```diff
@@ -217,4 +217,8 @@ public void testSyntheticCount() {
     public void testUnknownCount() {
         assertEquals(0, getVisitor().unknownCount, "unknownCount");
     }
+    @Test
```
Please add a blank line to separate methods (watch out for trailing whitespace).

PabloNicolasDiaz replied:
fixed in https://github.com/apache/commons-bcel/pull/290/commits/ed41dba669d77ce3a7fca96471691eef4af9a79d

garydgregory commented on src/main/java/org/apache/bcel/classfile/Record.java:
```diff
@@ -0,0 +1,152 @@
+            final int attributesCount = input.readUnsignedShort();
+            final Attribute[] attributes = new Attribute[attributesCount];
+            for (int j = 0 ; j< attributesCount; j++)
+                attributes[j] = Attribute.readAttribute(input, constantPool);
```
Always use `{` and `}`.

garydgregory reviewed (changes requested):
More comments.

garydgregory commented:
@PabloNicolasDiaz 
The build is broken with:
```
Error:  Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.18.5:cmp (default-cli) on project bcel: There is at least one incompatibility: org.apache.bcel.classfile.Visitor.visitRecordComponent(org.apache.bcel.classfile.RecordComponentInfo):METHOD_ADDED_TO_INTERFACE -> [Help 1]
```
Use a default method.
Run `mvn` by itself before you push to catch all build checks.

PabloNicolasDiaz commented:
Hello @garydgregory, let me figure out if i can solve all the errors and if RecordComponentInfo needs to be acceder to users. RecordComponentInfo should be mapped to every `components` element in a record attribute (it's documented in https://docs.oracle.com/javase/specs/jvms/se14/preview/specs/records-jvms.html#jvms-4.7.30).

garydgregory commented:
@PabloNicolasDiaz 
The PR still causes the build to fail.
Also, note that the code coverage is insufficient as reported by Codecov:
```
Attention: Patch coverage is 50.56180% with 44 lines in your changes are missing coverage. Please review.

Project coverage is 65.09%. Comparing base (f32c53c) to head (ce967c6).
Report is 35 commits behind head on master.

Files	Patch %	Lines
...rc/main/java/org/apache/bcel/classfile/Record.java	53.84%	16 Missing and 2 partials ⚠️
...org/apache/bcel/classfile/RecordComponentInfo.java	55.55%	11 Missing and 1 partial ⚠️
...a/org/apache/bcel/classfile/DescendingVisitor.java	0.00%	8 Missing ⚠️
...main/java/org/apache/bcel/classfile/JavaClass.java	66.66%	2 Missing and 2 partials ⚠️
...c/main/java/org/apache/bcel/classfile/Visitor.java	0.00%	2 Missing ⚠️Additional details and impacted files
@@             Coverage Diff              @@
##             master     #290      +/-   ##
============================================
- Coverage     65.10%   65.09%   -0.02%     
- Complexity     3893     3909      +16     
============================================
  Files           364      366       +2     
  Lines         15657    15745      +88     
  Branches       1956     1967      +11     
============================================
+ Hits          10194    10249      +55     
- Misses         4549     4577      +28     
- Partials        914      919       +5     
```
You can also use JaCoCo locally with `mvn clean site -P jacoco` and open the site in `target/site` where you'll find a nice and detailed HTML report; look for JaCoCo in the left-hand side menu.

garydgregory commented:
@PabloNicolasDiaz 
It does not look like you rebased on git master and run `mvn` locally because this fails for me:
```
git apply ~/Downloads/290.diff.txt
.../Downloads/290.diff.txt:402: trailing whitespace.

.../Downloads/290.diff.txt:445: trailing whitespace.
     * @throws IOException
.../Downloads/290.diff.txt:446: trailing whitespace.
     * @throws ClassFormatException
.../Downloads/290.diff.txt:455: trailing whitespace.

../Downloads/290.diff.txt:459: trailing whitespace.
```

garydgregory commented:
Hello @PabloNicolasDiaz 

Thank you for your update. The current code coverage is poor as reported by JaCoCo, for example, the new Record class is only at 66% and 58% for branches.
<img width="1096" alt="image" src="https://github.com/apache/commons-bcel/assets/1187639/6faf192c-8925-4f61-abb1-2038cd0b5fbe">

garydgregory commented:
@PabloNicolasDiaz 
Again, rebase on git master if you haven't, run `mvn` (or the same as `maven.yml`: `mvn --show-version --batch-mode --no-transfer-progress -DtrimStackTrace=false -Ddoclint=none`), and fix issues, starting with:

```
git apply ~/Downloads/290.patch
/Users/.../Downloads/290.patch:73: trailing whitespace.

/Users/.../Downloads/290.patch:80: trailing whitespace.

/Users/.../Downloads/290.patch:93: trailing whitespace.

/Users/garydgregory/Downloads/290.patch:124: trailing whitespace.

/Users/garydgregory/Downloads/290.patch:135: trailing whitespace.
```

PabloNicolasDiaz commented:
Hello @garydgregory, I've rebased and added a new test to check dump functionality and removed unnecessary methods. I've run the coverage report and its much better now

![image](https://github.com/apache/commons-bcel/assets/166959/b6aec97a-5d31-4819-afc7-dcbbec41bfa5)

Thanks for your time!

garydgregory commented:
@PabloNicolasDiaz 
For nth time, you MUST run `mvn` (no extra arguments needed) on the command line before you push and fix issues, otherwise you'll just keep on breaking PR builds :-(

garydgregory commented:
Hello @PabloNicolasDiaz 
Thank you for your update. Some new methods are still 0% tested, for example the new `accept` methods:
<img width="1102" alt="image" src="https://github.com/apache/commons-bcel/assets/1187639/1ce716dd-4551-4406-98ab-7e2f58720601">
and
<img width="1058" alt="image" src="https://github.com/apache/commons-bcel/assets/1187639/502a403f-8d4e-4501-8ad4-3e095709e5f1">

garydgregory commented on src/main/java/org/apache/bcel/classfile/RecordComponentInfo.java:
```diff
@@ -0,0 +1,102 @@
+        return index;
+    }
+
+    public int getDescriptorIndex() {
```
Hi @PabloNicolasDiaz 
This method is neither used or tested, how about removing until it is needed? 
Or, is it used by you in an other app?
It is also missing a Javadoc.

garydgregory commented on src/main/java/org/apache/bcel/classfile/Visitor.java:
```diff
@@ -207,6 +207,14 @@ default void visitNestMembers(final NestMembers obj) {
     }
 
+    /**
+     * @since 6.9.0
```
Hi @PabloNicolasDiaz 
The Javadoc is missing the first sentence.

garydgregory commented on src/main/java/org/apache/bcel/classfile/Visitor.java:
```diff
@@ -237,4 +245,14 @@ default void visitStackMapType(final StackMapType obj) {
+     * @since 6.9.0
+     */
+    default void visitRecordComponent(RecordComponentInfo obj) {
+
```
Remove blank line or add a comment that says `// noop`.

garydgregory commented on src/main/java/org/apache/bcel/classfile/JavaClass.java:
```diff
@@ -906,4 +911,27 @@ public String toString() {
+    /**
+     * Query method to check if this class was declared as a record
+     *
+     * @return true if a record attribute is present, false otherwise
```
Missing `@since` tag.

garydgregory commented on src/main/java/org/apache/bcel/classfile/Record.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    /**
+     * @return deep copy of this record attribute
```
Javadoc is missing its first sentence.

garydgregory commented on src/main/java/org/apache/bcel/classfile/Record.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    /**
+     * @return array of Record Component Info elements.
```
Javadoc is missing its first sentence.

garydgregory commented on src/main/java/org/apache/bcel/classfile/Record.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    /**
+     * @return String representation, i.e., a list of classes.
```
Javadoc is missing its first sentence.

garydgregory commented on src/main/java/org/apache/bcel/classfile/RecordComponentInfo.java:
```diff
@@ -0,0 +1,102 @@
+        this.constantPool = constantPool;
+    }
+
+    public void dump(DataOutputStream file) throws IOException {
```
Missing Javadoc

garydgregory commented on src/test/java/org/apache/bcel/classfile/RecordTestCase.java:
```diff
@@ -0,0 +1,107 @@
+public class RecordTestCase extends AbstractTestCase {
+
+    /**
+     * An record type, once compiled, should result in a class file that is marked such that we can determine from the
```
`An record` -> `A record`

garydgregory commented on src/test/java/org/apache/bcel/classfile/RecordTestCase.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    /**
+     * An simple record with two simple fields, an integer and a String field, should
```
`An simple` -> `A simple`.

garydgregory reviewed (changes requested):
Hi @PabloNicolasDiaz 
Thank you for your update. I think we're almost done, please see my remaining comments.

PabloNicolasDiaz commented on src/main/java/org/apache/bcel/classfile/RecordComponentInfo.java:
```diff
@@ -0,0 +1,102 @@
+        v.visitRecordComponent(this);
+    }
+
+    public int getIndex() {
```
I made the getters `getIndex()`, `getDescriptorIndex()` and `getAttributes()` to give users access to this data. I'm just pushing a commit right now  with the corresponding javadoc comment and covered them in `RecordTestCase`

garydgregory commented:
@markro49 
Would you please review this PR?

paulk-asert commented on src/test/java/org/apache/bcel/classfile/RecordTestCase.java:
```diff
@@ -0,0 +1,120 @@
+    @Test
+    public void testRecordClassSaysItIs() throws ClassNotFoundException, ClassFormatException, IOException {
+        final JavaClass clazz = new ClassParser("src/test/resources/record/SimpleRecord.class").parse();
+        assertTrue(clazz.isRecord(), "Expected SimpleEnum class to say it was a record - but it didn't !");
```
SimpleEnum -> SimpleRecord in test failure message

markro49 commented on src/main/java/org/apache/bcel/classfile/JavaClass.java:
```diff
@@ -906,4 +911,28 @@ public String toString() {
+                break;
+            }
+        }
+        this.computedRecord = false;
```
shouldn't this be set to true?

PabloNicolasDiaz replied:
fixed in https://github.com/apache/commons-bcel/pull/290/commits/e3c2f2eff322f998fd97416becb60c68d6a3ba48

markro49 commented on src/test/java/org/apache/bcel/classfile/RecordTestCase.java:
```diff
@@ -0,0 +1,120 @@
+     */
+    @Test
+    public void testRecordClassSaysItIs() throws ClassNotFoundException, ClassFormatException, IOException {
+        final JavaClass clazz = new ClassParser("src/test/resources/record/SimpleRecord.class").parse();
```
Should SimpleRecord.java be added to the resources/record directory for backup/documentation purposes?

PabloNicolasDiaz replied:
added in https://github.com/apache/commons-bcel/pull/290/commits/e3c2f2eff322f998fd97416becb60c68d6a3ba48

markro49 commented:
completed my review

garydgregory commented on src/main/java/org/apache/bcel/classfile/Visitor.java:
```diff
@@ -237,4 +249,15 @@ default void visitStackMapType(final StackMapType obj) {
+    /**
+     * Visits a {@link RecordComponentInfo} object.
+     *
+     * @param obj record component to visit
```
This causes:
```
[ERROR] /Users/garydgregory/git/commons-bcel/src/main/java/org/apache/bcel/classfile/Visitor.java:256: error: @param name not found
[ERROR]      * @param obj record component to visit
[ERROR]               ^
[ERROR] 1 error
[ERROR] 100 warnings
```
when I run `mvn clean clean install site -P jacoco`

garydgregory replied:
Please rebase on git master to see this error in the CI build. For some reason, we had doclint disabled.

garydgregory replied:
@PabloNicolasDiaz 
You still have not fixed the above, the build fails.

garydgregory commented:
@PabloNicolasDiaz 
The API `RecordComponentInfo#getConstantPool()` is neither used nor tested.

PabloNicolasDiaz commented:
Hi and sorry @garydgregory :(, I have run `mvn` but didn't see the error in this class. I've fixed it in my last commit 

Thanks for your time and sorry again
